coredump: add config/API for registering memory for dump

mrkhldn opened the pull request:
Introduce new config and API to allow for dynamically adding
(and removing) regions of memory to be collected at time of coredump.

Add coredump_memory test to verify registered memory is collected in
dump when configured.

Signed-off-by: Mark Holden <mholden@fb.com>

dcpleung commented on include/debug/coredump.h:
```diff
@@ -136,6 +150,9 @@ void coredump_buffer_output(uint8_t *buf, size_t buflen);
 int coredump_query(enum coredump_query_id query_id, void *arg);
 int coredump_cmd(enum coredump_cmd_id cmd_id, void *arg);
 
+void coredump_register_memory_region(struct coredump_mem_region_node_t *region);
+bool coredump_unregister_memory_region(struct coredump_mem_region_node_t *region);
```
I think this should hide behind the newly introduced kconfig. So that it will generate errors during compilation if the kconfig is not enabled, instead of silently ignoring the calls.

mrkhldn commented:
The pushes after [dcpleung](https://github.com/dcpleung)'s review have just been to rebase and make sure the change still merges/functions ok, no code changes.

tejlmand commented on tests/subsys/debug/coredump_memory/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(hello_world)
```
`hello_world` ?

tejlmand commented on doc/guides/debug_tools/coredump.rst:
```diff
@@ -28,6 +28,10 @@ Here are the choices regarding memory dump:
   dump is desired.
 
+* ``DEBUG_COREDUMP_MEMORY_DUMP_SELECT``: dumps the bare minimum memory content
+  that would be included with ``DEBUG_COREDUMP_MEMORY_DUMP_MIN`` plus any
+  additional memory regions that were registered for collection.
```
I think the documentation lacks description of how to register additional memory regions to be dumped.

How are users expected to know how to use this feature ?

I believe some references to `coredump_register_memory_region()` / API is a bare minimum here.

Commit:
coredump: add config/API for registering memory for dump

Introduce new config and API to allow for dynamically adding
(and removing) regions of memory to be collected at time of coredump.

Add coredump_memory test to verify registered memory is collected in
dump when configured.

Signed-off-by: Mark Holden <mholden@fb.com>

carlescufi commented:
@tejlmand please take another look

gmarull commented on include/debug/coredump.h:
```diff
@@ -161,6 +161,39 @@ void coredump_buffer_output(uint8_t *buf, size_t buflen);
+
+#include <sys/slist.h>
+
+/* Coredump memory region node */
+struct coredump_mem_region_node_t {
+	/* Node of single-linked list */
+	sys_snode_t	node;
+
+	/* Address of start of memory region */
+	uintptr_t	start;
+
+	/* Address of end of memory region */
+	uintptr_t	end;
+};
```
nit: please use Doxygen docstrings, `/**`.

gmarull commented on include/debug/coredump.h:
```diff
@@ -161,6 +161,39 @@ void coredump_buffer_output(uint8_t *buf, size_t buflen);
+ * @brief Unregister memory region for collection in dump
+ *
+ * @param region Struct containing start and end address of memory to no longer be collected
+ * @return True if region was removed, otherwise false
```
```suggestion
 * @retval true If region was removed
 * @retval false If region was not removed
```

mrkhldn commented:
Resolving merge conflicts

mrkhldn commented:
@tejlmand, @gmarull do the latest changes look ok?

gmarull reviewed (commented):
docs lgtm, will let maintainers in the area approve

cfriedt commented:
Would love to see a coredump pseudo-device with DT bindings 😀

cfriedt commented:
@mrkhldn  - This would be the **perfect** candidate for a pseudo-device driver + DT bindings 👀

mrkhldn commented:
> @mrkhldn - This would be the **perfect** candidate for a pseudo-device driver + DT bindings 👀

@cfriedt, this is an interesting idea. Can you help me understand the thinking a bit more? What advantage do we get changing this into a pseudo-device driver?

@dcpleung, any thoughts on the suggested approach?

dcpleung commented:
Hm... maybe the idea is to use device tree to define all the mappings at build time, and skip the runtime stuff? From a security standpoint this is (arguably) better since the region list can reside in read-only memory, and the list can't be manipulated externally to dump other addresses.

cfriedt commented:
> > @mrkhldn - This would be the **perfect** candidate for a pseudo-device driver + DT bindings 👀
> 
> @cfriedt, this is an interesting idea. Can you help me understand the thinking a bit more? What advantage do we get changing this into a pseudo-device driver?
> 
> @dcpleung, any thoughts on the suggested approach?

@mrkhldn, as @dcpleung mentioned, with a pseudo-device driver, addresses are stored in ROM as constants. I still think it's important to support the dynamic use case, of course.

However, a kind of nice part  about it is that you get multi-instance for free. Additional coredump areas can be specified with complete independence and flexibility for dedicated memory areas in heterogenous layouts (or potentially also DMA memory, or shared bus memory, or really any logged or buffered data). This type of information is fairly helpful in post-mortem analysis.
